File: Graphs/all_paths_source_target.py
# https://leetcode.com/problems/all-paths-from-source-to-target/
# Version 2 - Recursive
# Time complexity       O(n)
# Space complexity       O(longest path) -> O(n) in the worst case
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        self.graph = graph
        self.target = len(graph) - 1
        return self.dfs(0, [])
        
    def dfs(self, node, path):
        path.append(node)
        if node == self.target:
            return [path]
        return [child_path for child in self.graph[node] for child_path in self.dfs(child, path.copy()) if child_path]

# Version 1 - Iterative BFS
# Time complexity       O(n)
# Space Complexity      O(n)
from collections import deque
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        output = []
        q = deque([ [0] ])
        while q:
            path = q.popleft()
            if path[-1] == n - 1:
                output.append(path)
                continue
            for next_node in graph[path[-1]]:
                q.append(path + [next_node])

        return output

#### Key understandings ####
# When to add to a path is an important decision
# node.empty() doesn't necessarily mean that you've reached the end
# Once you have a working solution think of ones that require less copying

File: Graphs/allPathsSourceTarget.cpp
// Version 2 - DFS Improved
// Time complexity      O(n)
// Space complexity      O(n)
class Solution
{
public:
    void dfs(int node, vector<vector<int>> &graph, vector<int> &path, vector<vector<int>> &output)
    {
        path.push_back(node);
        if (node == graph.size() - 1) output.push_back(path);
        else
        {
            for (auto &next : graph[node])
            {
                dfs(next, graph, path, output);
            }
        }
        path.pop_back();
    }

    vector<vector<int>> allPathsSourceTarget(vector<vector<int>> &graph)
    {
        vector<vector<int>> output;
        vector<int> path;
        dfs(0, graph, path, output);
        return output;
    }
};

// Version 1 - DFS
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<int> path{0};
        return dfs(0, path, graph);
    }

    vector<vector<int>> dfs(int node, vector<int> path, vector<vector<int>>& graph)
    {
        vector<vector<int>> output;
        if (node == graph.size() - 1)
        {
            output.push_back(path);
        }
        else
        {
            for (auto& nextNode : graph[node] )
            {
                vector<int> pathIncNextNode(path);
                pathIncNextNode.push_back(nextNode);
                for (auto& childPath : dfs(nextNode, pathIncNextNode, graph))
                {
                    output.push_back(childPath);
                }
            }
        }
        return output
    }
};



// For edges connected to current node
// Append that edge to our current path
// Then call dfs on that node, giving it that path